### ISSUE TITLE
Missing sys/prctl.h

### ISSUE THREAD

**Issue by wolfviking0** [emscripten-core/emscripten#18141]
Please include the following in your bug report:

**Version of emscripten/emsdk:**
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.24 (68a9f990429e0bcfb63b1cde68bad792554350a5)
clang version 16.0.0 (https://github.com/llvm/llvm-project 277c382760bf9575cfa2eac73d5ad1db91466d3f)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /home/anthony.liot/devtools/emsdk/upstream/bin

**Failing command line in full:**
Missing header sys/prctl.h using cmake toolchain to build a library

I was looking but it seems the header cannot be found in the sysroot, should I modify code to be emscripten compatible ?

**Comment by wolfviking0** [comment]
I just wanted to add a comment, the same library was build few month ago on OSX using emsdk 3.1.17, 
The current version after brew install is 3.1.24 this version seems not working for me for this library

**Comment by sbc100** [comment]
`prctl.h` is a linux-specific header.  Emscripten doesn't support the `prctl` function that it defined, and it also doesn't include the header file (at least not as of #17704)

**Comment by sbc100** [comment]
Normally there is some way in the code to disable the use of linux-specific stuff like prctl.  Does the codebase you are working with not have such a thing?

**Comment by wolfviking0** [comment]
Hi @sbc100 

Right the prctl.h is linux only but my main problem is the same library version was working and building fine with emscripten 3.1.17 with no warning or error regarding the header.

Moving to a more recent version of emsdk this error appear I was suspecting something change in the cmake toolchain.

**Comment by sbc100** [comment]
Yes, we used to ship these linux-specific header even though we didn't support the functions they defined.

They were removed in #17704.  There was a changelog entry added in 3.1.20: https://github.com/emscripten-core/emscripten/blob/ea82d2d0fd6b7caef8223040e0f64222e241db44/ChangeLog.md?plain=1#L117-L119.  The hope was that this would cause minimum disruption and lead to more correct programs, but there will be some users who need to update their codebases in response.

**Comment by sbc100** [comment]
I guess in retrospect we could have softened the breakage but keeping the headers around with `#error` in them for a few releases.

**Comment by aisnote** [comment]
@sbc100  I meet this kind of issue too.  Now one component is build with 3.1.27 and another is built with 3.1.20, my question is : **are they compatible with each other?**   thanks.

The structure is like:
A(built with 3.1.20), B(built with 3.1.27), C(built with 3.1.20).

A load B with dlopen, and A load C with static linking.

**Comment by sbc100** [comment]
No, we do not (yet) make any guarantees that libraries built with one SDK version can be loaded by a main module built with a another SDK version.   One reason this might not work is if libc or lib++ change.. since these libraries are always built into the main module.  Another reason is changes to the JS library support code.

**Comment by aisnote** [comment]
so, for this:
> prctl.h is a linux-specific header. Emscripten doesn't support the prctl function that it defined

in my own's code, to use the same emsdk version, what I can do is just remove the prctl related code?

or any chance the emsdk to add back this prctl just for some user's building?

**Comment by sbc100** [comment]
@aisnote yes you are best to simply `ifdef` out the calls to `prctl`.   They will not work under emscripten.   If its your own code this should be fairly straight forward.. or am i missing something?

**Comment by aisnote** [comment]
This does make sense. Thanks to @sbc100.

**Comment by Ujjwal21** [comment]
@sbc100 I am also facing similar issue and using `ifdef` is not an option for me since I am using chromium code. Although I tried copying these missing header files from **earlier emsdk version** to latest under this path `emsdk/upstream/emscripten/system/include`. Even then it is not able to find the sys/prctl.h. Is that the correct way to go about it?

**Comment by sbc100** [comment]
@Ujjwal21 what chromium code are you trying to compile?   Can you link to it?  Are you sure you can't patch it?   If the code is using `sys/prctl.h` (and the functions defined there-in) I doubt it will function correctly when build for emscripten.  Did it previously work for you?

If you really want to create a `prctl.h` I would suggest starting by simple creating a an empty file to see if that works.  The right place to create it would be in `emsdk/upstream/emscripten/cache/sysroot/include/`

**Comment by Ujjwal21** [comment]
@sbc100 here is [link](https://source.chromium.org/chromium/chromium/src/+/main:third_party/perfetto/src/profiling/memory/client.cc?q=sys%2Fprctl.h&ss=chromium) to just one of the chromium files, we are using in our project that uses `sys/prctl.h`. There are many such other files like this in chromium codebase.

Yes, it used to work with earlier version of emscripten (fastcomp-1.38.30), and this version used to have all these system libraries.

I also tried pasting at this location -` emsdk/upstream/emscripten/cache/sysroot/include/`, but no luck :(

**Comment by sbc100** [comment]
The code you link to there doesn't look like it is designed to compiler on anything except android.   See https://source.chromium.org/chromium/chromium/src/+/main:third_party/perfetto/src/profiling/memory/README.md.   I dont think there is any way that could would work under emscripten.

If you really want to add `sys/prctl.h` it would need to go in `emsdk/upstream/emscripten/cache/sysroot/include/sys/prctl.h` (did you include the `sys` component when you tried to add it?)

**Comment by sbc100** [comment]
What parts of chromium are you hoping to build with emscripten?  Where are you planning on running that code?  i.e.are you targeting the web?
